### PR TITLE
feat: amino acids are now stored

### LINKED ISSUE #85: Include Amino Acid Sequences
Currently the database only stores nucleotide sequences. The database should also store amino acid sequences.

Details:
- Probably create subtasks and link them here to split this into smaller bits.

### PR DESCRIPTION
**Pull request by Taepper**
resolve #85

### CODE REVIEW

File: src/silo/storage/sequence_store.cpp [hunk 1]
```diff
@@ -104,7 +104,7 @@ void silo::SequenceStorePartition::fillNBitmaps(const std::vector<std::string>&
    tbb::blocked_range<unsigned> const range(0, genomes.size());
    tbb::parallel_for(range, [&](const decltype(range)& local) {
       /// For every symbol, calculate all sequence IDs that have that symbol at that position
-      std::vector<unsigned> positions_with_nucleotide_symbol_n(SYMBOL_COUNT);
```
Comment (Taepper) [inline src/silo/storage/sequence_store.cpp:107]: Note the change in the test cases is due to this bugfix: I accidentally always inserted `SYMBOL_COUNT` zeros into the `positions_with_nucleotide_symbol_n`, making the position `0` of all sequences contain the symbol `N`

File: src/silo/storage/aa_store.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+   const size_t genome_length = positions.size();
+
+   aa_symbol_x_bitmaps.resize(sequence_count + sequences.size());
+
+   tbb::blocked_range<unsigned> const range(0, sequences.size());
+   tbb::parallel_for(range, [&](const decltype(range)& local) {
+      /// For every symbol, calculate all sequence IDs that have that symbol at that position
+      std::vector<unsigned> positions_with_aa_symbol_x;
+      for (unsigned genome = local.begin(); genome != local.end(); ++genome) {
+         for (unsigned pos = 0, limit2 = genome_length; pos != limit2; ++pos) {
```
Comment (fengelniederhammer) [inline src/silo/storage/aa_store.cpp:82]:
```suggestion
      for (unsigned sequence_id = local.begin(); sequence_id != local.end(); ++sequence_id) {
         for (unsigned position = 0; position != genome_length; ++position) {
```

Comment (fengelniederhammer) [reply src/silo/storage/aa_store.cpp:82]: Is `position != genome_length` actually any more efficient than `position < genome_length`? `<` would make the purpose of the statement clearer.

File: src/silo/prepare_dataset.cpp [hunk 3]
```diff
@@ -239,6 +239,19 @@ void silo::partitionData(
...
+      FastaReader sequence_input(sequence_filename);
+
+      std::filesystem::path aa_folder = output_folder;
+      aa_folder += "gene_" + aa_name + std::filesystem::path::preferred_separator;
+
+      create_directory(aa_folder);
+
+      partitionSequenceFile(sequence_input, aa_folder, chunk_names, key_to_chunk, reference_genome);
+   }
+
```
Comment (fengelniederhammer) [inline src/silo/prepare_dataset.cpp:254]: This is very similar to the previous for loop. How easy would it be to extract a function?

File: src/silo/prepare_dataset.cpp [hunk 4]
```diff
@@ -398,6 +411,18 @@ void silo::sortChunks(
...
+         input_filename += "gene_" + aa_name + std::filesystem::path::preferred_separator;
+         input_filename += silo::buildChunkString(chunk.part, chunk.chunk) + ZSTDFASTA_EXTENSION;
+         sequence_inputs.emplace_back(input_filename, reference_sequence);
+
+         std::filesystem::path output_filename = output_folder;
+         output_filename += "gene_" + aa_name + std::filesystem::path::preferred_separator;
+         create_directory(output_filename);
+         output_filename += silo::buildChunkString(chunk.part, chunk.chunk) + ZSTDFASTA_EXTENSION;
+         sequence_outputs.emplace_back(output_filename, reference_sequence);
+      }
```
Comment (fengelniederhammer) [inline src/silo/prepare_dataset.cpp:425]: Same here

**Review by fengelniederhammer — APPROVED** [review]

File: src/silo/prepare_dataset.cpp (hunk 4 above)
Comment (Taepper) [reply src/silo/prepare_dataset.cpp:425]: Must of the code is managing the filenames and creating the directory. I will tackle that this is at most two lines of code in the body in #156, which should make it idiomatic enough?

File: src/silo/storage/aa_store.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+void silo::AAStorePartition::fillIndexes(const std::vector<std::string>& sequences) {
+   const size_t genome_length = positions.size();
+   static constexpr int COUNT_SYMBOLS_PER_PROCESSOR = 64;
+   tbb::blocked_range<unsigned> const range(
+      0, genome_length, genome_length / COUNT_SYMBOLS_PER_PROCESSOR
+   );
+   tbb::parallel_for(range, [&](const decltype(range)& local) {
+      /// For every symbol, calculate all sequence IDs that have that symbol at that position
```
[the remark this answers is not in the document]
Comment (Taepper) [reply src/silo/storage/aa_store.cpp]: Unfortunately I couldn't as I didn't want to introduce additional buffer allocations. Although I remove the comment and gave the buffer a more descriptive name

**Pull request merged**